samza-api/src/main/java/org/apache/samza/operators/MessageStream.java

```diff
@@ -16,173 +16,109 @@
 
 
 /**
- * This class defines either the input or output streams to/from the operators. Users use the API methods defined here to
- * directly program the stream processing stages that processes a stream and generate another one.
- *
- * @param <M>  Type of message in this stream
+ * Represents a stream of values of type {@link Message}. A {@link MessageStream} can be transformed into another
+ * {@link MessageStream} by applying the transforms in this API.
+ * @param <M>  type of {@link Message}s in this stream
  */
 @InterfaceStability.Unstable
-public class MessageStream<M extends Message> {
-
-  private final Set<Operator> subscribers = new HashSet<>();
-
-  /**
-   * Helper method to get the corresponding list of subscribers to a specific {@link MessageStream}.
-   *
-   * NOTE: This is purely an internal API and should not be used directly by programmers.
-   *
-   * @return A unmodifiable set containing all {@link Operator}s that subscribe to this {@link MessageStream} object
-   */
-  public Collection<Operator> getSubscribers() {
-    return Collections.unmodifiableSet(this.subscribers);
-  }
-
-  /**
-   * Public API methods start here
-   */
+public interface MessageStream<M extends Message> {
 
   /**
-   * Defines a function API that takes three input parameters w/ types {@code A}, {@code B}, and {@code C} and w/o a return value
+   * Applies the 1:1 {@link MapFunction} to this {@link MessageStream} and returns the resulting {@link MessageStream}.
    *
-   * @param <A>  the type of input {@code a}
-   * @param <B>  the type of input {@code b}
-   * @param <C>  the type of input {@code c}
-   */
-  @FunctionalInterface
-  public interface VoidFunction3<A, B, C> {
-    public void apply(A a, B b, C c);
-  }
-
-  /**
-   * Method to apply a map function (1:1) on a {@link MessageStream}
-   *
-   * @param mapper  the mapper function to map one input {@link Message} to one output {@link Message}
+   * @param mapFn  the function to map one input {@link Message} to its corresponding output {@link Message}
    * @param <OM>  the type of the output {@link Message} in the output {@link MessageStream}
-   * @return the output {@link MessageStream} by applying the map function on the input {@link MessageStream}
+   * @return the output {@link MessageStream} after applying the map function on the input {@link MessageStream}
    */
-  public <OM extends Message> MessageStream<OM> map(Function<M, OM> mapper) {
-    Operator<OM> op = Operators.<M, OM>getStreamOperator(m -> new ArrayList<OM>() { {
-        OM r = mapper.apply(m);
-        if (r != null) {
-          this.add(r);
-        }
-      } });
-    this.subscribers.add(op);
-    return op.getOutputStream();
-  }
+  <OM extends Message> MessageStream<OM> map(MapFunction<M, OM> mapFn);
 
   /**
-   * Method to apply a flatMap function (1:n) on a {@link MessageStream}
+   * Applies the 1:n {@link FlatMapFunction} to {@link Message} in this {@link MessageStream}, concatenates the results,
```

Comment:
nits:

1. /s/the 1:n/the provided/
2. /s/to {@link Message} in this {@link MessageStream}/on this {@link MessageStream}/. Is n't it implicit that applying a function on a stream applies it to every message on the stream? For example, the param docs state the function applied on the stream (instead of each message on it).